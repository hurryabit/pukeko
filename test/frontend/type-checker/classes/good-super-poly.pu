class C a where
  c :: a -> Int
class (C a) <= D a where
  d :: a -> Int
data T a = | C a
instance cInt :: C Int where
  c x = x
instance dInt :: D Int where
  d x = x
instance cT :: (C a) => C (T a) where
  c t =
    match t with
    | C x -> c x
instance dT :: (D a) => D (T a) where
  d t =
    match t with
    | C x -> d x
f :: (D a) => T a -> Int
f x = c x
