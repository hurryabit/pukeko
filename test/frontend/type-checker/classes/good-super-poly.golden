data C a =
       | .C (a -> Int)
c :: forall a. C a -> a -> Int =
  \@a (dict :: C a) ->
    case dict of
    | .C c -> c
data D a =
       | .D (C a) (a -> Int)
D.C :: forall a. D a -> C a =
  \@a (dict :: D a) ->
    case dict of
    | .D D.C _ -> D.C
d :: forall a. D a -> a -> Int =
  \@a (dict :: D a) ->
    case dict of
    | .D _ d -> d
data T a =
       | C a
cInt :: C Int = .C @Int cInt.c
cInt.c :: Int -> Int = \(x :: Int) -> x
dInt :: D Int = .D @Int dInt.C dInt.d
dInt.C :: C Int = cInt
dInt.d :: Int -> Int = \(x :: Int) -> x
cT :: forall a. C a -> C (T a) =
  \@a (c.a :: C a) -> .C @(T a) (cT.c @a c.a)
cT.c :: forall a. C a -> T a -> Int =
  \@a (c.a :: C a) (t :: T a) ->
    case t of
    | C x -> c @a c.a x
dT :: forall a. D a -> D (T a) =
  \@a (d.a :: D a) -> .D @(T a) (dT.C @a d.a) (dT.d @a d.a)
dT.C :: forall a. D a -> C (T a) =
  \@a (d.a :: D a) -> cT @a (D.C @a d.a)
dT.d :: forall a. D a -> T a -> Int =
  \@a (d.a :: D a) (t :: T a) ->
    case t of
    | C x -> d @a d.a x
f :: forall a. D a -> T a -> Int =
  \@a (d.a :: D a) (x :: T a) -> c @(T a) (cT @a (D.C @a d.a)) x
